app/com/linkedin/drelephant/tuning/obt/TuningTypeManagerOBT.java

```diff
@@ -0,0 +1,408 @@
+package com.linkedin.drelephant.tuning.obt;
+
+import com.avaje.ebean.Expr;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.JsonNodeFactory;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.linkedin.drelephant.ElephantContext;
+import com.linkedin.drelephant.tuning.AbstractTuningTypeManager;
+import com.linkedin.drelephant.tuning.JobTuningInfo;
+import com.linkedin.drelephant.tuning.Particle;
+import com.linkedin.drelephant.tuning.ExecutionEngine;
+import com.linkedin.drelephant.tuning.TuningHelper;
+import controllers.AutoTuningMetricsController;
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import models.AppResult;
+import models.JobDefinition;
+import models.JobExecution;
+import models.JobSavedState;
+import models.JobSuggestedParamSet;
+import models.JobSuggestedParamValue;
+import models.TuningAlgorithm;
+import models.TuningJobDefinition;
+import models.TuningParameter;
+import org.apache.commons.io.FileUtils;
+import org.apache.log4j.Logger;
+import play.libs.Json;
+import org.apache.hadoop.conf.Configuration;
+
+
+public abstract class TuningTypeManagerOBT extends AbstractTuningTypeManager {
+
+  private static final String PARAMS_TO_TUNE_FIELD_NAME = "parametersToTune";
+  private static final String PYTHON_PATH_CONF = "python.path";
+  private static final String PSO_DIR_PATH_ENV_VARIABLE = "PSO_DIR_PATH";
+  private static final String PYTHON_PATH_ENV_VARIABLE = "PYTHONPATH";
+  private String PYTHON_PATH = null;
+  private String TUNING_SCRIPT_PATH = null;
+  private final Logger logger = Logger.getLogger(getClass());
+
+
+  /*
+    Intialize any prequisite require for Optimizer
+    Calls once in lifetime of the flow
+   */
+  public abstract void intializePrerequisite(TuningAlgorithm tuningAlgorithm,
+      JobSuggestedParamSet jobSuggestedParamSet);
+
+  /*
+    Optimize search space
+    call after each execution of flow
+   */
+  public abstract void parameterOptimizer(List<AppResult> appResults, JobExecution jobExecution);
+
+  /**
+   * Swarm size specific to OBT
+   * @return
+   */
+  protected abstract int getSwarmSize();
+
+  public TuningTypeManagerOBT() {
+    tuningType = "OBT";
+    Configuration configuration = ElephantContext.instance().getAutoTuningConf();
+
+    PYTHON_PATH = configuration.get(PYTHON_PATH_CONF);
```

Comment:
AbstractTuningType is the one developer should extend to have new tuning type  .  Since PSO & IPSO have lot of code common which have no relation  with HBT , we have the class TuningTypeManagerOBT. If developers write for new tuning type  , he/she should extend AbstractTuningType

Reply:
Here I am not talking about adding a new tuning type but adding an OBT algorithm. In my comments when I am referring to an algorithm, it is OBT algorithm. 
Now, considering OBT as a separate component, algorithm implementation should be abstracted out and OBT class should invoke specific algorithm implementations so that if we add another OBT algorithm in future, that will be done by implementing an interface.
TuningTypeManager is a top level interface and OBT algorithms are specific to OBT. So TuningTypeManagerOBT should not contain algorithm specific code related to PSO, IPSO,etc. I see a PSO script being invoked here for instance.

Reply:
done